codex/sales/slotqueue.nim

```diff
@@ -0,0 +1,257 @@
+          maxWorkers = DefaultMaxWorkers,
+          maxSize = DefaultMaxSize): SlotQueue =
+
+  let mworkers = if maxWorkers == 0'u: DefaultMaxWorkers
+                 elif maxWorkers > maxSize: maxSize
+                 else: maxWorkers
+
+  SlotQueue(
+    # Add 1 to always allow for an extra item to be pushed onto the queue
+    # temporarily. After push (and sort), the bottom-most item will be deleted
+    queue: newAsyncHeapQueue[SlotQueueItem](maxSize.int + 1),
+    maxWorkers: mworkers,
+    workers: newSeqOfCap[SlotQueueWorker](mworkers),
+    running: false
+  )
+
+proc new*(_: type SlotQueueWorker): SlotQueueWorker =
+  SlotQueueWorker(
+    processing: newFuture[void]("slotqueue.worker.processing")
+  )
+
+proc init*(_: type SlotQueueItem,
+          requestId: RequestId,
+          slotIndex: uint64,
+          ask: StorageAsk,
+          expiry: UInt256): SlotQueueItem =
+
+  SlotQueueItem(
+    requestId: requestId,
+    slotIndex: slotIndex,
+    ask: ask,
+    expiry: expiry
+  )
+
+proc init*(_: type SlotQueueItem,
+           request: StorageRequest,
+           slotIndex: uint64): SlotQueueItem =
+
+  SlotQueueItem.init(request.id, slotIndex, request.ask, request.expiry)
+
+proc init*(_: type SlotQueueItem,
+          requestId: RequestId,
+          ask: StorageAsk,
+          expiry: UInt256): seq[SlotQueueItem] =
+
+  var i = 0'u64
+  proc initSlotQueueItem: SlotQueueItem =
+    let sqi = SlotQueueItem.init(requestId, i, ask, expiry)
+    inc i
+    return sqi
+
+  var sqis = newSeqWith[uint64](ask.slots, initSlotQueueItem())
+  Rng.instance.shuffle(sqis)
+  return sqis
+
+proc init*(_: type SlotQueueItem,
+          request: StorageRequest): seq[SlotQueueItem] =
+
+  return SlotQueueItem.init(request.id, request.ask, request.expiry)
+
+proc requestId*(self: SlotQueueItem): RequestId = self.requestId
+
+proc slotIndex*(self: SlotQueueItem): uint64 = self.slotIndex
+
+proc running*(self: SlotQueue): bool = self.running
+
+proc len*(self: SlotQueue): int = self.queue.len
+
+proc `$`*(self: SlotQueue): string = $self.queue
+
+proc `onProcessSlot=`*(self: SlotQueue, onProcessSlot: OnProcessSlot) =
+  self.onProcessSlot = some onProcessSlot
+
+proc activeWorkers*(self: SlotQueue): int = self.workers.len
+
+proc contains*(self: SlotQueue, sqi: SlotQueueItem): bool =
+  self.queue.contains(sqi)
+
+proc pop*(self: SlotQueue): Future[?!SlotQueueItem] {.async.} =
+  try:
+    let sqi = await self.queue.pop()
+    return success(sqi)
+  except CatchableError as err:
+    return failure(err)
+
+proc push*(self: SlotQueue, sqi: SlotQueueItem): ?!void =
+  if self.contains(sqi):
+    let err = newException(SlotQueueItemExistsError, "item already exists")
+    return failure(err)
+
+  if err =? self.queue.pushNoWait(sqi).mapFailure.errorOption:
+    return failure(err)
+
+  if self.queue.full():
+    # delete the last item
+    self.queue.del(self.queue.size - 1)
+
+  doAssert self.queue.len <= self.queue.size
```

Comment:
The `AsyncHeapQueue` will raise an error if pushing past the max size, whereas `SlotQueue` will not. `SlotQueue` does this by setting the underlying `AsyncHeapQueue` size to `maxSize + 1` to ensure that items can always be added, and once items are added, the last item is deleted if the queue is full.  Because `SlotQueue` adds this functionality on top of `AsyncHeapQueue`, we should ensure `SlotQueue` checks that once an item is pushed, it should never have more than `self.queue.size - 1` items. I did [update the `doAssert` check](https://github.com/codex-storage/nim-codex/pull/455/commits/6d3033a10321539c30359046f20f3ba42c9ae51c) to add the `.size - 1` (previously it was just `doAssert self.queue.len <= self.queue.size`).